### PR TITLE
changefeedccl: mark changefeeds idle when no recent kv activity

### LINKED ISSUE #74749: changefeed: Mark job idle
Add infrastructure to keep track of the number of messages emitted by aggregators, and communicate
this information to the frontier processor.

The job may decide to mark itself idle (depends on #74747) when:
  * the number of emitted messages remains at 0 for some period of (configurable) time
  * It is not currently performing a backfill
  * It has just written a checkpoint record
  * The checkpoint includes PTS record (protect when idle)
  
Edit CRDB-8671

### PR DESCRIPTION
**Pull request by samiskin**
Changefeeds run indefinitely, which poses a problem for serverless in
being able to shut down nodes when there isn't any activity going on.
This change makes it so that a changefeed that hasn't emitted any KVs in
the last `changefeed.idle_timeout` duration (default 10 minutes) will be
marked as Idle and be able to be shut down.  Any new activity will
result in the changefeed no longer being idle.

Fixes #74749

Release note: None

### CODE REVIEW

**Comment by samiskin** [comment]
This PR is currently based on top of https://github.com/cockroachdb/cockroach/pull/75897, only the last commit is new

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, 2 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @samiskin)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1324 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75166#-MuvftwOee7DVJRp5kCx-r2-1324:-Mv0B62vD7pCsBk12wcZ:b-a7kpn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ea810c6e30f9c721b77925d1363e317a4a3a242/pkg/ccl/changefeedccl/changefeed_processors.go#L1324)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

If this happens should I fall back to attempting to parse as a resolvedspan?  I'm concerned about the case where the clusterversion updates but there's messages from the old version in flight.
</blockquote></details>

Not possible.  If version is active, it means that everybody emits new proto; and if not active, everybody emits old one.

I would actually consider dropping AggregatorProgressUpdate message type.  Can we just do away with emitting
jobspb.ResolvedSpans?  it's a list of resolved spans right now, but we can add stats related fields in there.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75166#-Mv0D-uh1ku-xVTwV83P:-Mv0D-uh1ku-xVTwV83Q:bj3qufd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ea810c6e30f9c721b77925d1363e317a4a3a242/pkg/ccl/changefeedccl/changefeed_processors.go#L1409)):*
> ```Go
> func (cf *changeFrontier) maybeMarkJobIdle(recentKVCount uint64) {
> 	versionGate := cf.flowCtx.Cfg.Settings.Version.IsActive(cf.Ctx, clusterversion.ChangefeedIdleness)
> 	if !versionGate || cf.spec.JobID == 0 {
> ```

the version gate check happens in multiple places.  Consider initializing some boolean flags once.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75166#-Mv0EF7C-bUdzTUf6h7n:-Mv0EF7C-bUdzTUf6h7o:b-zemy45) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ea810c6e30f9c721b77925d1363e317a4a3a242/pkg/ccl/changefeedccl/changefeed_processors.go#L1430)):*
> ```Go
> 
> 	// The timestamp should be protected during the idle periods since nodes
> 	// running idle changefeeds can be freely shut down
> ```

I think this is okay... BUT, I want to pull out this particular change into separate PR.
We now unconditionally lay PTS records, and we advance them with certain rate.
We were planning to do this anyway (and not just for this usecase).  I think we should
get the PTS specific change in first.  You could move this into another commit in this PR; 
I just think it's a separate logical change.
___
*[pkg/ccl/changefeedccl/helpers_test.go, line 331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75166#-Mv0AYDu9H3hw18GiXzr:-Mv0AYDu9H3hw18GiXzs:b-mwmjgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ea810c6e30f9c721b77925d1363e317a4a3a242/pkg/ccl/changefeedccl/helpers_test.go#L331)):*
> ```Go
> 		DistSQL:          &execinfra.TestingKnobs{Changefeed: &TestingKnobs{}},
> 		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
> 		Server:           &server.TestingKnobs{},
> ```

why do we need this now?

File: pkg/ccl/changefeedccl/changefeed_processors.go [hunk 1]
```diff
@@ -1269,42 +1300,63 @@ func (cf *changeFrontier) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetad
...
+
+	var resolvedSpans jobspb.ResolvedSpans
+	if cf.flowCtx.Cfg.Settings.Version.IsActive(cf.Ctx, clusterversion.ChangefeedIdleness) {
+		if err := protoutil.Unmarshal([]byte(*raw), &resolvedSpans); err != nil {
+			return errors.NewAssertionErrorWithWrappedErrf(err,
+				`unmarshalling aggregator progress update: %x`, raw)
+		}
+
+		cf.maybeMarkJobIdle(resolvedSpans.Stats.RecentKvCount)
+	} else {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_processors.go:1328]: Add todo to remove this in 22.2

File: pkg/ccl/changefeedccl/changefeedbase/settings.go [hunk 2]
```diff
@@ -59,6 +59,16 @@ var SlowSpanLogThreshold = settings.RegisterDurationSetting(
 	settings.NonNegativeDuration,
 )
 
+// IdleTimeout controls how long the changefeed will wait for a new KV being
+// emitted before marking itself as idle
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeedbase/settings.go:63]: nit: period@end of sentence.

**Review by miretskiy — APPROVED** [review]
You are likely missing "make generate" state since changing cluster version also changes some of the generated files (plus new setting).  
Other than that, good job!  We should let serverless know about new setting.

**Pull request merged**